File: ios/NexusCore/Sources/NexusCore/Models/NightQueueWorkshop.swift
import Foundation

/// Readiness workshop (baker-internal#111) — the daylight half of the night
/// queue, as the backend proxies it from the partner adapter. Mirror of the
/// workshop types in `src/frontend/src/api.ts`; wire keys are snake_case, so
/// the shared `.nexusREST` decoder applies.
///
/// Three reads, one conversation, and one write. The write — arming — posts a
/// readiness comment and mints the `night-queue` label, which is the moment an
/// issue is handed to an unattended agent for a whole night. Every guard on it
/// lives in the adapter; the types here exist so the phone can MIRROR those
/// refusals rather than discover them after a round trip.
///
/// Nothing in this file re-states the readiness bar. The criteria are served
/// by `/api/night-queue/readiness` precisely so a Swift copy cannot drift from
/// what the 01:00 planner actually enforces.

// MARK: - Candidates

/// Why an open PR blocks an issue: an explicit "Fixes #N", one of our own
/// `nq/` branches, or a branch named after the issue.
public enum BlockerReason: String, Decodable, Sendable {
    case linked
    case nqBranch = "nq_branch"
    case branchName = "branch_name"
    case unknown

    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = BlockerReason(rawValue: raw) ?? .unknown
    }
}

/// An open PR that already implements the issue.
public struct NightQueueBlocker: Decodable, Sendable, Equatable {
    public let number: Int
    public let url: String
    public let branch: String
    public let reason: BlockerReason

    public init(number: Int, url: String, branch: String, reason: BlockerReason) {
        self.number = number
        self.url = url
        self.branch = branch
        self.reason = reason
    }
}

/// What structurally stops an issue being armed. `nil` on the wire means
/// nothing does — which is NOT the same as "ready": nothing in the candidate
/// list has been judged against the bar until it is assessed.
public enum BlockedReason: String, Decodable, Sendable, Equatable {
    /// Standing policy: the unattended agent never touches its own runtime or
    /// the surface it reports through.
    case excluded
    /// Already labelled — it is in tonight's queue.
    case queued
    case openPr = "open_pr"
    case unknown

    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = BlockedReason(rawValue: raw) ?? .unknown
    }
}

/// An open issue that COULD be armed, carrying the reason it could not.
///
/// Blocked candidates are returned rather than filtered out, and the UI shows
/// them greyed with their reason: a list that silently omits an issue teaches
/// you it does not exist, while "PR #212 already implements this" teaches you
/// where the work went.
public struct NightQueueCandidate: Decodable, Sendable, Identifiable, Hashable {
    public var id: String { "\(repo)#\(number)" }
    public let repo: String
    public let number: Int
    public let title: String
    public let url: String
    public let updatedAt: String?
    public let updatedTs: Int?
    public let labels: [String]
    public let queued: Bool
    public let excluded: Bool
    public let openPr: NightQueueBlocker?
    /// `nil` = nothing structurally blocks arming. Covers excluded / queued /
    /// open_pr in one field, so a gate written against it cannot miss a case
    /// the way a hand-rolled disjunction did on desktop.
    public let blocked: BlockedReason?

    public static func == (lhs: NightQueueCandidate, rhs: NightQueueCandidate) -> Bool {
        lhs.id == rhs.id
    }

    public func hash(into hasher: inout Hasher) { hasher.combine(id) }
}

public struct NightQueueCandidatesResponse: Decodable, Sendable {
    /// False when the backend has no assistant URL/key configured.
    public let configured: Bool?
    public let candidates: [NightQueueCandidate]
    /// Count with no structural blocker. Deliberately NOT called "armable" —
    /// nothing here has been judged against the bar yet.
    public let unblocked: Int?
    public let generatedAt: Int?
    public let cached: Bool?
    /// True when the list below is the last good answer, not a current one.
    public let stale: Bool?
    public let error: String?
}

// MARK: - The bar

public struct ReadinessCriterion: Decodable, Sendable, Identifiable {
    public let id: String
    public let label: String
    public let requirement: String
    /// Set when the criterion only applies to some issues (reachability).
    public let conditional: String?
}

/// The readiness bar, served rather than duplicated. `night-queue/readiness.py`
/// is the single definition the 01:00 planner composes its prompt from, so a
/// workshop that blessed issues against a second copy would teach Keith to
/// trust a green light that does not bind.
public struct ReadinessResponse: Decodable, Sendable {
    public let configured: Bool?
    public let criteria: [ReadinessCriterion]
    /// The verbatim bar the planner enforces.
    public let barText: String?
    public let commentTemplate: String?
    public let excludedRepos: [String]?
    public let error: String?
}

// MARK: - Assessment

public enum CriterionStatus: String, Decodable, Sendable {
    case met, missing, na, unknown

    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = CriterionStatus(rawValue: raw) ?? .unknown
    }
}

public struct AssessedCriterion: Decodable, Sendable, Identifiable {
    public let id: String
    public let label: String
    public let status: CriterionStatus
    public let note: String
}

/// One issue judged against the bar. Costs a model call, so it is requested
/// per-issue and on demand — and a failure is an error rather than an empty
/// verdict: "I could not judge this" must never render as "nothing is wrong".
public struct AssessmentResponse: Decodable, Sendable {
    public let repo: String
    public let number: Int
    public let title: String
    public let url: String
    public let state: String
    public let labels: [String]
    public let queued: Bool
    public let excluded: Bool
    public let openPr: NightQueueBlocker?
    /// Recomputed from the criteria, never taken from the model.
    public let ready: Bool
    /// False when the verdict was decided without a model call (excluded repo).
    public let assessed: Bool
    public let summary: String
    public let criteria: [AssessedCriterion]
    /// Gaps arrive as `<TODO: …>` and must be resolved before arming.
    public let draftComment: String

    /// Whether this verdict is about `candidate`. The stale-assessment guard:
    /// an assessment takes tens of seconds, so a verdict can land after the
    /// reader has moved on, and arming on a mismatched pair would post one
    /// issue's spec onto another and label it.
    public func describes(_ candidate: NightQueueCandidate) -> Bool {
        repo == candidate.repo && number == candidate.number
    }
}

// MARK: - The conversation before arming

/// A Partner conversation opened about one issue (baker-internal#131).
///
/// The adapter composes the seed — the readiness bar plus the issue text inside
/// its fence — and holds it against the session, so it rides on every turn
/// rather than only the first. Nothing here builds that fence: a second
/// spelling in Swift would be an injection path into the arming decision the
/// conversation exists to inform.
///
/// The session is the adapter's, so the app adopts it before chatting. The
/// Partner drafts and argues; it cannot arm, and its seed says so.
public struct DiscussResponse: Decodable, Sendable {
    /// The adapter-side session to adopt.
    public let sessionId: String
    public let repo: String
    public let number: Int
    public let title: String
    public let url: String
    public let sessionTitle: String
}

// MARK: - Arming (the one write)

public struct ArmDecision: Decodable, Sendable {
    /// Autonomy-ledger class, e.g. `night-queue-arm:selfie-wall`. The
    /// `night-queue` prefix is permanently denylisted from ever going auto.
    public let armClass: String?
    public let recorded: Bool?
    public let promotable: Bool?
    public let streak: Int?
    /// Audit-only failure: reported ALONGSIDE the success, never over it, so a
    /// ledger hiccup cannot un-arm a correctly armed issue.
    public let error: String?

    private enum CodingKeys: String, CodingKey {
        // `class` is a Swift keyword. `.convertFromSnakeCase` leaves a key
        // with no underscores alone, so this matches the wire key verbatim.
        case armClass = "class"
        case recorded, promotable, streak, error
    }
}

public struct ArmResponse: Decodable, Sendable {
    public let repo: String
    public let number: Int
    public let title: String
    public let url: String
    public let queued: Bool
    public let label: String
    public let commentPosted: Bool
    public let decision: ArmDecision?
}

// MARK: - The arm gate

/// Whether Arm may be offered at all, and if not, why.
///
/// This mirrors refusals the adapter enforces anyway. It lives here, in
/// NexusCore, rather than inside a SwiftUI view for two reasons: it is the
/// safety-critical half of the workshop and deserves tests, and the view
/// should never be the only place a refusal is written down.
///
/// It deliberately does NOT re-implement every adapter rule — the bookkeeping
/// -prefix check stays server-side, where the planner's own filter is defined.
/// Two answers to one question would leave this side untested against a real
/// night.
public enum ArmGate {
    /// The adapter's own floor for "long enough to be a spec".
    public static let minimumCommentCharacters = 40
    /// The marker the assessor uses for a gap it will not decide for you.
    public static let unresolvedMarker = "<TODO"

    public enum Refusal: Equatable, Sendable {
        /// Nothing has been judged yet — the bar costs a model call per issue.
        case notAssessed
        /// The verdict on screen is about a DIFFERENT issue. Bug #1 on desktop:
        /// switching issues mid-flight landed verdict A under heading B, and
        /// Arm took the repo from B with the comment from A.
        case verdictIsForAnotherIssue
        /// Bug #2 on desktop: the UI showed a loud "PR #212 already implements
        /// this" banner and then let you arm anyway.
        case blocked(BlockedReason)
        case unresolvedTodo
        case tooShort
        /// Already armed in this session — the comment is posted, the label minted.
        case alreadyArmed

        /// Shown on the disabled control, so the reason is never a mystery.
        public var reason: String {
            switch self {
            case .notAssessed:
                return "Assess it against the bar first."
            case .verdictIsForAnotherIssue:
                return "This verdict is about another issue — assess this one."
            case .blocked(.openPr):
                return "A PR already implements this."
            case .blocked(.excluded):
                return "This repo never runs unattended by standing policy."
            case .blocked(.queued):
                return "Already labelled — it is in tonight's queue."
            case .blocked:
                return "Blocked — see above."
            case .unresolvedTodo:
                return "Resolve the <TODO: …> first."
            case .tooShort:
                return "Too short to be a spec — the adapter needs \(minimumCommentCharacters) characters."
            case .alreadyArmed:
                return "Already armed."
            }
        }
    }

    /// `nil` means Arm may be offered. Order matters only for which reason the
    /// reader is shown; every refusal here is also enforced by the adapter.
    public static func evaluate(
        candidate: NightQueueCandidate,
        assessment: AssessmentResponse?,
        draft: String,
        armed: Bool = false
    ) -> Refusal? {
        if armed { return .alreadyArmed }
        // `blocked` covers excluded / queued / open_pr in one. Checked before
        // the verdict so a blocked issue reads as blocked rather than unjudged.
        if let blocked = candidate.blocked { return .blocked(blocked) }
        guard let assessment else { return .notAssessed }
        guard assessment.describes(candidate) else { return .verdictIsForAnotherIssue }
        let text = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.contains(unresolvedMarker) { return .unresolvedTodo }
        if text.count < minimumCommentCharacters { return .tooShort }
        return nil
    }
}

File: src/backend/routes/night-queue-workshop.ts
import { FastifyInstance } from 'fastify';
import { loadConfig, resolveAssistantKey, resolveEnvVars } from '../config.js';
import { createHermesClient, type HermesFetch } from '../hermes/client.js';
import type { NexusConfig } from '@nexus/shared';

/** The adapter answers errors as FastAPI's `{"detail": "..."}`. Unwrap it so
 * the card shows "this issue is already queued" rather than raw JSON. */
function detailOf(err: any, fallback: string): string {
  const raw: string = err?.message || fallback;
  try {
    const parsed = JSON.parse(raw);
    if (parsed && typeof parsed.detail === 'string') return parsed.detail;
  } catch {
    /* not JSON — the message is already human-readable */
  }
  return raw;
}

interface WorkshopRoutesOptions {
  fetchImpl?: HermesFetch;
}

/**
 * Readiness workshop (baker-internal#111) — the daylight half of the night
 * queue. Three reads, one conversation, and one write.
 *
 * The write, `POST /api/night-queue/arm`, is the moment an issue joins
 * tonight's queue for an unattended agent. It is a pure passthrough on
 * purpose: every guard (excluded repos, unresolved `<TODO:`, a spec the
 * planner would filter as its own bookkeeping, comment-before-label ordering,
 * the autonomy ledger) lives in the partner adapter, where the trust machinery
 * already is. Re-implementing any of it here would give two answers to one
 * question, and this side would be the one nobody tested against a real night.
 */
export function createWorkshopRoutes(
  load: () => NexusConfig = loadConfig,
  options: WorkshopRoutesOptions = {},
) {
  return async function registerWorkshopRoutes(fastify: FastifyInstance) {
    const client = () => {
      const config = load();
      const url = resolveEnvVars(config.assistant.url || '').trim();
      const key = resolveAssistantKey(config);
      if (!url || !key) return undefined;
      return createHermesClient({ url, key, fetchImpl: options.fetchImpl });
    };

    // Reads: fail-soft, matching /api/routines and /api/night-queue.
    fastify.get('/api/night-queue/readiness', async () => {
      const hermes = client();
      if (!hermes) return { configured: false, criteria: [] };
      try {
        return { configured: true, ...(await hermes.nightQueueReadiness() as object) };
      } catch (err: any) {
        return { configured: true, criteria: [], error: err?.message || 'Readiness fetch failed.' };
      }
    });

    fastify.get('/api/night-queue/candidates', async () => {
      const hermes = client();
      if (!hermes) return { configured: false, candidates: [] };
      try {
        return { configured: true, ...(await hermes.nightQueueCandidates() as object) };
      } catch (err: any) {
        return { configured: true, candidates: [], error: err?.message || 'Candidate fetch failed.' };
      }
    });

    // Assessment costs a model call, so it is a POST the user triggers, and a
    // failure is an error rather than an empty verdict: "I could not judge
    // this" must never render as "nothing is wrong with it".
    fastify.post('/api/night-queue/assess', async (request, reply) => {
      const { repo, number } = (request.body ?? {}) as { repo?: string; number?: number };
      if (!repo || typeof number !== 'number') {
        reply.code(400);
        return { error: 'repo (string) and number (number) are required.' };
      }
      const hermes = client();
      if (!hermes) {
        reply.code(400);
        return { error: 'Assistant URL and key must be configured in Settings.' };
      }
      try {
        return await hermes.assessIssue(repo, number);
      } catch (err: any) {
        // The client attaches the upstream status (see hermes/client.ts) —
        // 404 for an unknown issue, 502 for an unreadable verdict. Use it
        // rather than sniffing the message.
        reply.code(err?.status ?? 502);
        return { error: detailOf(err, 'Assessment failed.') };
      }
    });

    // Opening a conversation about an issue. It creates an assistant session
    // and writes nothing to GitHub, so it is not gated like the arm below —
    // but it is a POST because it has an effect, and the adapter's 403 for an
    // excluded repo reaches the client unchanged.
    //
    // The seed (the readiness bar plus the issue text inside its fence) is
    // composed by the adapter, never here and never by the client: a second
    // spelling of that fence would be an injection path into the very decision
    // the conversation exists to inform.
    fastify.post('/api/night-queue/discuss', async (request, reply) => {
      const { repo, number, draft } = (request.body ?? {}) as {
        repo?: string; number?: number; draft?: string;
      };
      if (!repo || typeof number !== 'number') {
        reply.code(400);
        return { error: 'repo (string) and number (number) are required.' };
      }
      const hermes = client();
      if (!hermes) {
        reply.code(400);
        return { error: 'Assistant URL and key must be configured in Settings.' };
      }
      try {
        return await hermes.discussIssue({
          repo, number,
          draft: typeof draft === 'string' ? draft : undefined,
        });
      } catch (err: any) {
        reply.code(err?.status ?? 502);
        return { error: detailOf(err, 'Could not open the conversation.') };
      }
    });

    // Getting the agreed wording back out of that conversation
    // (baker-internal#132). A read, and the whole point of it is that nothing
    // between the transcript and the draft field is allowed to be creative:
    // the adapter returns the last fenced block verbatim or says it found
    // none. The card puts the result in the EDITABLE draft, where every arm
    // guard still applies to it.
    //
    // `:sessionId` is the ADAPTER session id from `discuss`, not the nexus id
    // the chat endpoints are keyed by — the card holds both.
    fastify.get('/api/night-queue/discuss/:sessionId/comment', async (request, reply) => {
      const { sessionId } = request.params as { sessionId: string };
      const hermes = client();
      if (!hermes) {
        reply.code(400);
        return { error: 'Assistant URL and key must be configured in Settings.' };
      }
      try {
        return await hermes.discussComment(sessionId);
      } catch (err: any) {
        // 404 is "no workshop conversation with that id" and reads quite
        // differently from "the adapter is down". Keep them apart.
        reply.code(err?.status ?? 502);
        return { error: detailOf(err, 'Could not read the comment.') };
      }
    });

    // The write. The adapter's status reaches the card unchanged, because the
    // card branches on it: 403 is standing policy, 409 is already queued or
    // closed, 400 is a spec the adapter refused. Flattening them all to 502
    // would turn four different conversations into one shrug.
    fastify.post('/api/night-queue/arm', async (request, reply) => {
      const { repo, number, comment, decided_by } = (request.body ?? {}) as {
        repo?: string; number?: number; comment?: string; decided_by?: string;
      };
      if (!repo || typeof number !== 'number' || typeof comment !== 'string') {
        reply.code(400);
        return { error: 'repo, number and comment are required.' };
      }
      const hermes = client();
      if (!hermes) {
        reply.code(400);
        return { error: 'Assistant URL and key must be configured in Settings.' };
      }
      try {
        // Forwarded so the autonomy ledger records WHERE the decision was
        // made — a phone tap and a desk session are the same write but not the
        // same act. Defaults to the desktop's value when a client omits it.
        return await hermes.armIssue({
          repo, number, comment,
          decided_by: typeof decided_by === 'string' && decided_by.trim()
            ? decided_by.trim().slice(0, 64)
            : 'nexus-workshop',
        });
      } catch (err: any) {
        reply.code(err?.status ?? 502);
        return { error: detailOf(err, 'Arming failed.') };
      }
    });
  };
}

export const registerWorkshopRoutes = createWorkshopRoutes();
